hcipy/wavefront_sensing/pyramid.py

```diff
@@ -62,30 +62,38 @@ class PyramidWavefrontSensorOptics(WavefrontSensorOptics):
 	----------
 	input_grid : Grid
 		The grid on which the input wavefront is defined.
+	output_grid : Grid
+		The grid on which the output wavefront is defined.
 	separation : scalar
 		The separation between the pupils. The default takes the input grid extent as separation.
+	D : scalar
+		The diameter of the aperture.
 	wavelength_0 : scalar
 		The reference wavelength that determines the physical scales.
 	q : scalar
 		The focal plane oversampling coefficient. The default uses the minimal required sampling.
+	num_airy : scalar
+		The radius of the focal plane spatial filter in units of lambda/D at the reference wavelength.
 	refractive_index : callable
 		A callable that returns the refractive index as function of wavelength.
 		The default is a refractive index of 1.5.
-	num_airy : scalar
-		The radius of the focal plane spatial filter in units of lambda/D at the reference wavelength.
 	'''
-	def __init__(self, input_grid, separation=None, wavelength_0=1, q=None, num_airy=None, refractive_index=lambda x: 1.5):
+	def __init__(self, input_grid, output_grid, separation=None, D=None, wavelength_0=1, q=None, num_airy=None, refractive_index=lambda x: 1.5):
 		if not input_grid.is_regular:
 			raise ValueError('The input grid must be a regular grid.')
 
 		self.input_grid = input_grid
-		D = np.max(input_grid.delta * (input_grid.shape - 1))
+		self.output_grid = output_grid
+
+		if D is None:
+			D = np.max(input_grid.delta * (input_grid.shape - 1))
 
 		if separation is None:
 			separation = D
 
+		# Create the intermediate focal grid
 		# Oversampling necessary to see all frequencies in the output wavefront sensor plane
-		qmin = max(2 * separation / D, 1)
+		qmin = np.ceil(max(output_grid.x.ptp() / input_grid.x.ptp(), 2))
```

Comment:
The extent of a regular grid is not `x.ptp()` but `grid.delta * grid.dims`. What about elongated pupils?